Add copy constructor for Position so that it behaves correctly in std co...

matthewlai opened the pull request:
The rule of 3 in C++ says we should have all 3 of - destructor, copy constructor, and assignment operator, if we have any of the 3.

Our design of the Position class is a little strange and there's not really a sensible way to implement a dtor.

However, we can implement the copy ctor using assignment operator, and that makes it behave correctly in std containers.

I discovered this working on my own stuff based on Stockfish. Not sure if the default copy ctor is currently used anywhere in Stockfish, but even if it doesn't fix a bug right now, it may prevent a hard-to-find bug in the future.

If it's not used right now, it will just be optimized out by the compiler.

mcostalba commented:
And why we should have all 3 ?

Sorry to ask, I'd just want to figure out if there is some kind of
understanding here of we are just facing another case of "driven by
mnemonics".



Position in a container? What we are talking about?

BTW Position is good as is. Copy c'tor is disabled now and it should remain
like this.

matthewlai commented:
Between the copy ctor and the assignment operator, if one of them is required, that means the class needs to be deep-copied, meaning the other one is probably required as well (since the default implementation would be wrong). In this case the default copy ctor is wrong.

The destructor is USUALLY also required since the usual reason for requiring a deep copy is there are resources that live outside of the object and is managed by the object (eg, on the heap). In our case the destructor is not required because StateInfo is either managed by the caller, or lives inside the object.

The "rule" is a memory aid, yes, but I did think through the rationale behind it and see that it applies in this case before suggesting it.

The copy ctor is NOT currently disabled. Defining constructors that take arguments only disables the default (no argument) constructor. Not the copy ctor.

Actually disabling it is one option, but does that make more sense than implementing the copy constructor and making it copyable?

The Position class represents a chess position, and is conceptually self contained. Is there a reason why it should NOT be copyable? If disabling the copy constructor and adding one that works are about the same amount of work and add the same amount of complexity, why not make it copyable?

If you don't want it to be copyable, doesn't it make sense to make it not assign-able as well? From a design point of view, why should it be assignable but not copyable?

Stockfish doesn't currently put Positions in containers. I am modifying Stockfish for my research where I do, and the fact that there is a wrong copy ctor resulted in a very hard to find bug. I just thought I would contribute this patch back because this behaviour of the Position class is clearly unintended, and may result in a hard to find bug in the future.

mcostalba commented:
Copy c'tor was properly disabled already many years ago:

https://github.com/official-stockfish/Stockfish/commit/a042449f5d63ad3459f92f93

Actually it managed to sneak in again. But I think makes more sense to
disable it again because:

1) Is not used

2) In case someone uses it, then he should be well aware of its side
effects. Disabiling it is a kind of warning.


And to answer to your comments. No, Position is not self contained:

Position p = pos;

pos = p;

// pos != p

So disabling it makes clear that the copy construct does not really _copy_
the position.

mcostalba commented:
So, becuase of:

Position p = pos;


assert(!memcmp(&p, &pos, sizeof(Position)));


then it is better to disable the copy'ctor.

matthewlai commented:
I guess it's a difference between our understandings of "self-contained", then.

I am speaking of "self-containment" in a conceptual sense, so I am of the opinion that the correct solution is to implement operator!= and operator== so that in your example, pos == p, because they represent the same position. That said, this is getting into the design territory, and is not as clear cut.

I would also consider std::vector to be self-contained, even though it also fails your memory test.

It conceptually represents a self-contained collection of elements, and uses operator overloading, constructors, and destructor, to behave like a self-contained collection. It's the same with std::string.

I like this self-contained version of Position better, and I don't see any downside besides a few extra lines of code with no performance penalty if unused, so I'll be implementing the copy ctor, operator==, and operator!= in my copy. Thanks for the idea. Which way Stockfish goes is obviously up to you, as it's entirely a design decision at this point.

If at some point Stockfish needs to keep track of and manipulate a large number of positions - benchmarking, testing, tuning, for example, I'd think the self-contained version would be much more convenient to use.

mcostalba commented:
I understand your point: if you need it you want it.

But I'd prefer people to think twice before copying a Position, that's why disabling it rings a bell. Otherwise people naively uses the provided copy 'ctor and then pass hours to find out why their version crashes.



For the other people reading here and possibly not aware of the big fun that is C++

Position p = pos;

It _seems_ the assignment operator, but it is actually the copy c'tor ! As can be verified with below test code.

  struct Test {
    Test() {}
    Test(const Test&) { std::cout << "I'm the copy c'tor" << std::endl;}
    void operator=(const Test&)  { std::cout << "I'm the assignment" << std::endl;}
  };

void main()
{
    Test t1;

    Test t2 = t1; // It will be printed "I'm the copy c'tor"
}

matthewlai commented:
Definitely. Leaving the default copy constructor in there is a bad idea.

The question is between disabling it and implementing one that does the right thing.

If the copy constructor and comparison operators are implemented, the Position class should behave intuitively no matter what people do with it (copy, assign, compare, and using it in containers), and there wouldn't be any hidden gotchas.

Copying is expensive of course, but people are generally aware that copying a large object is expensive.

mcostalba commented:
The problem is the copy ctor can't do the right thing. State pointer is
anyhow detached and user should be aware of the consequences:  hiding this
fact to the user is a poor choice IMO, so better disable it and if someone
needs it he will implement what it is OK for him, knowing what he is doing
and the limitations.





On Thursday, January 1, 2015, matthewlai <notifications@github.com> wrote:

> Definitely. Leaving the default copy constructor in there is a bad idea.
>
> The question is between disabling it and implementing one that does the
> right thing.
>
> If the copy constructor and comparison operators are implemented, the
> Position class should behave intuitively no matter what people do with it
> (copy, assign, compare, and using it in containers), and there wouldn't be
> any hidden gotchas.
>
> Copying is expensive of course, but people are generally aware that
> copying a large object is expensive.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/183#issuecomment-68493264>
> .
>

zamar commented:
As Marco said the most sensible thing is to disable the copy constructor. If we can't do something properly, then it's best not to do it at all...

mcostalba commented:
Matthew, if you agree, could you please update your pull request putting
the copy c'tor before 'public', as in the patch I pointed out? So that
possibly it will be committed. Thanks.


On Fri, Jan 2, 2015 at 9:53 AM, Joona Kiiski <notifications@github.com>
wrote:

> As Marco said the most sensible thing is to disable the copy constructor.
> If we can't do something properly, then it's best not to do it at all...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/183#issuecomment-68514232>
> .
>

matthewlai commented:
Updated.

I can't say I agree with the decision, especially when there is already a "Position(const Position& pos, Thread* th)" constructor that is essentially a copy constructor, implemented like in the original patch. I would think having the copy constructor do the same thing as both that constructor and the assignment operator makes things more consistent. Otherwise it could be argued that operator= making a copy of StateInfo hides it from the user as well, and people may use it without thinking twice.

But of course this is not my project to make design decisions on :).

glinscott commented:
Ok, change looks good!

zamar commented:
approved.
